mtp_transaction_uploader/upload.py

```diff
@@ -0,0 +1,125 @@
+    new_dates = []
+    with Connection(settings.SFTP_HOST, username=settings.SFTP_USER,
+                    private_key=settings.SFTP_PRIVATE_KEY) as conn:
+        with conn.cd(settings.SFTP_DIR):
+            dir_listing = conn.listdir()
+            for filename in dir_listing:
+                m = file_pattern.match(filename)
+
+                if m:
+                    date = datetime.strptime(m.group(2), '%d%m%y')
+                    if date > last_date:
+                        local_path = os.path.join(settings.DS_NEW_FILES_DIR,
+                                                  filename)
+                        new_files.append(local_path)
+                        new_dates.append(date)
+                        conn.get(filename, local_path=local_path)
+
+    # find last dated file
+    if len(new_dates) > 0:
+        new_last_date = sorted(new_dates)[-1]
+
+    FileDownload = namedtuple('FileDownload', ['new_last_date', 'new_files'])
+    return FileDownload(new_last_date, new_files)
+
+
+def upload_transactions_from_files(files):
+    transactions = []
+    for filename in files:
+        transactions += get_transactions_from_file(filename)
+    conn = get_authenticated_connection()
+    conn.bank_admin.transactions.post(transactions)
+
+
+def get_transactions_from_file(filename):
+    with open(filename) as f:
+        data_services_file = parse_file(f)
+
+    if not data_services_file.is_valid():
+        raise ValueError("%s invalid: %s" % (filename, data_services_file.errors))
+
+    transactions = []
+    for record in data_services_file:
+        if (record.transaction_code == TransactionCode.credit_bacs_credit or
+                record.transaction_code == TransactionCode.credit_sundry_credit):
+            transaction = {}
+            transaction['amount'] = record.amount/100
+            transaction['sender_sort_code'] = record.originators_sort_code
+            transaction['sender_account_number'] = record.originators_account_number
+            transaction['sender_name'] = record.transaction_description
+            transaction['reference'] = record.reference_number
+            transaction['received_at'] = record.date
+
+            parsed_ref = parse_reference(record.reference_number)
+            if parsed_ref:
+                number, dob = parsed_ref
+                transaction['prisoner_number'] = number
+                transaction['prisoner_dob'] = parsed_ref
+
+            transactions.append(transaction)
+    return transactions
+
+
+def parse_reference(ref):
+    m = ref_pattern.match(ref)
+
+    if m:
+        date_str = '%s/%s/%s' % (m.group(2), m.group(3), m.group(4))
+        try:
+            dob = datetime.strptime(date_str, '%d/%m/%Y')
+        except ValueError:
+            dob = datetime.strptime(date_str, '%d/%m/%y')
+
+            # set correct century for 2 digit year
+            if dob.year > datetime.today().year - 10:
+                dob = dob.replace(year=dob.year-100)
```

Comment:
This is a neat hack :+1:

That said… are we 100% sure this system won't need to deal with under-10s? @tonyduarte?